lib/kino/download.ex

```diff
@@ -0,0 +1,79 @@
+defmodule Kino.Download do
+  @moduledoc false
+
+  use Kino.JS
+  use Kino.JS.Live
+
+  @doc """
+  Renders a download button.  The content of downloaded file is
+  prepared on the server by invoking the `content_fun` argument only
+  when the button is clicked.
+
+  ## Examples
+
+      Kino.Download.new("file.json", fn ->
+        ~s/{"foo": "bar"}/
+      end)
+
+  """
+  def new(filename, content_fun) do
+    Kino.JS.Live.new(__MODULE__, {filename, content_fun})
+  end
+
+  @impl true
+  @spec init({any, any}, Kino.JS.Live.Context.t()) :: {:ok, Kino.JS.Live.Context.t()}
+  def init({filename, content_fun}, ctx) do
+    {:ok, assign(ctx, filename: filename, content_fun: content_fun)}
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    {:ok, Map.take(ctx.assigns, [:filename]), ctx}
+  end
+
+  @impl true
+  def handle_event("prepare_download", %{"filename" => filename}, ctx) do
+    ctx = assign(ctx, filename: filename)
+
+    # To keep things generic the content is a base64-encoded binary
+    file_content =
+      ctx.assigns.content_fun.()
+      |> Base.encode64()
```

Comment:
```suggestion
    file_content = ctx.assigns.content_fun.()
```

We don't want to encode on Elixir side, because then the payload becomes bigger. With {:binary, %{}, file_content}` we can send an arbitrary Elixir binary, and it is received as ArrayBuffer in JS. You can then convert it to base64, as we do [here](https://github.com/jonatanklosko/kino_lab/blob/450c02fe34549637c2c87b5de46fe196a77832cf/lib/assets/camera_processor/main.js#L119-L129).